### ISSUE TITLE
allow StreamJsonRpc to raise cancellation when caller side close the connection

### ISSUE THREAD

**Issue by heejaechang** [Microsoft/vs-streamjsonrpc#64]
I see that, if I raise cancellation and close Rpc connection right away, sometimes, cancellation message doesn't reach server side before Rpc connection is disconnected. making server side not know the operation is actually cancelled.

for us, this is bad since we are waiting for other connection to happen inside of service hub, and hang there since it doesn't know the operation is actually cancelled.

I think StreamJsonRpc should either guarantee of ordering (cancellation and disconnect) or give a way to configure streamJsonRpc to always raise cancellation if connection is disconnected.

**Comment by heejaechang** [comment]
tagging @IlyaBiryukov and @AArnott

**Comment by IlyaBiryukov** [comment]
The workaround is to subscribe to server's JsonRpc.Disconnected event and cancel the server methods when the disconnection happens

**Comment by heejaechang** [comment]
that's true, but that makes one to always create linked cancellation token. one from the given one, and one from the disconnect method.

**Comment by AArnott** [comment]
> I think StreamJsonRpc should either guarantee of ordering (cancellation and disconnect) 

There is no way StreamJsonRpc can "guarantee ordering" since a disconnection can happen at any time.

> or give a way to configure streamJsonRpc to always raise cancellation if connection is disconnected.

This seems reasonable. Our existing `Disconnected` event predated our `CancellationToken` support. And not all server methods may even accept a `CancellationToken`. So if we do offer a way to automatically cancel in-progress operations when the connection drops, it will of course only impact those methods that take `CancellationToken` while other methods will still have to listen for the `Disconnected` event or add `CancellationToken` to their method signature, which would be a non-breaking change for the JSON-RPC clients.

**Comment by AArnott** [comment]
@heejaechang I take it this is the root cause of your servicehub hosting process spinning the CPU endlessly? If so, would you use this new feature of StreamJsonRpc if we fixed it fast enough? Which release would you check the fix into, so we can make the feature addition to the right branch of this project as well?

**Comment by heejaechang** [comment]
I am thinking bring it to 15.5 preview1. right now, I implemented what @IlyaBiryukov suggested.

**Comment by AArnott** [comment]
Is that workaround you refer to hooking the `JsonRpc.Disconnected` event? If so, we'll keep this issue open to consider fixing in a future version, but without worrying about helping with your fix in the short term. That sounds like a good call to me.

**Comment by heejaechang** [comment]
@AArnott I think it is good for short term, but it will be nice if we can remove it from outside since LUT and Razor is also using our OOP for roslyn solution and they also need to do what I am doing to get proper cancellation support.

**Comment by AArnott** [comment]
Sounds reasonable.

Next step then is to make a proposal for how we expose this option. A few ideas come to mind:

1. an assembly attribute you define in your assembly
1. a property you set on a `JsonRpc` instance
1. a parameter passed to the `JsonRpc` ctor.

I'm liking the property approach myself.

Also, while I think I wouldn't have minded if the default behavior originally was to cancel requests on Disconnect, now that that is _not_ the default behavior, it could be a breaking change if we were to change the default behavior. We can poll our partners if there is any interest in making the cancel-on-disconnect be the default behavior. Thoughts?

**Comment by heejaechang** [comment]
I think cancel on disconnect would be good default behavior since, I think, that would align with what people would expect. "I disposed JsonRpc and I expect remote side goes away". 

I think "I disconnect and I want my remote side to keep doing whatever it was doing even though I completely lost control of that service" will be more rare case.

**Comment by heejaechang** [comment]
2 seems easier non breaking way to add new functionality. 3 seems more discoverable. but 2 and 3 seem not exclusive to each other. you could just add both? anyway, 2 seems fine as well.
